### PR TITLE
[data, docs] fix: avoid long-context batching cold-start gate

### PR DESCRIPTION
**Pull request by YLlllllllllll**
### What does this PR do?

Adds an opt-in context-aware candidate-buffer policy for long-context,
main-process dynamic batching:

- 512K, MBS1: minimum 24 candidates per DP rank
- 1M, MBS1: minimum 24 candidates per DP rank

The default remains `fixed / 200`, so existing jobs, including 128K jobs, do
not change. The opt-in policy supports any DP size; it still fails closed for
worker runtime, effective-token counting, non-text data, MBS>1, and context
lengths other than 512K/1M.

The value 24 is a minimum readiness gate, not a capacity. A rank keeps fetching
past 24 when the buffered token count is below one batching window.

### Root cause

`TextBatchingStrategy` starts packing only after both conditions are true:

1. `len(buffer) >= dyn_bsz_buffer_size`
2. buffered tokens cover the batching token budget

With long samples, enough tokens can arrive far before the fixed
200-candidate gate opens. The first micro-batch then spends most of its time
fetching and transforming candidates that are not selected by the ordered
greedy packer.

Natural-completion DP1 measurements were:

| Token budget | Buffer 200 | Buffer 24 | Source fill at buffer 24 |
| --- | ---: | ---: | ---: |
| 512K | 5861.24 s | 576.75 s | 99.7480% |
| 1M | 5063.80 s | 589.63 s | 99.6184% |

These are CPU-only first-micro-batch data-ready times, not first-forward or
steady-step times.

### DP scaling and shuffle-seed evidence

The probe retained the production stream, transform, dynamic batching,
packing, collator, and SP slice, but did not build a model or run an
accelerator collective. One process represented each independent DP stream.

Global source fill is:

```text
sum(selected_source_tokens) / (dp_size * max_seq_len)
```

For 1M/minimum-24:

| Shuffle seed | DP2 | DP4 | DP8 |
| --- | ---: | ---: | ---: |
| 42 | 99.5129% | 98.6507% | 98.7175% |
| 43 | 98.5127% | 99.4050% | 98.7576% |

All six points pass the 98.5% global source-fill gate.

For 512K/seed-43:

| DP | Buffer | Ready p50 | Global source fill |
| ---: | ---: | ---: | ---: |
| 1 | 8 | 191.61 s | 93.1940% |
| 1 | 24 | 478.76 s | 98.7411% |
| 2 | 8 | 244.67 s | 95.5143% |
| 2 | 16 | 392.61 s | 98.2879% |
| 2 | 24 | 567.99 s | 98.2879% |
| 4 | 16 | 351.15 s | 97.9949% |
| 4 | 24 | 547.36 s | 98.4475% |
| 4 | 32 | 696.47 s | 98.4475% |
| 8 | 32 | 667.50 s | 98.6668% |

DP2 buffers 16/24 and DP4 buffers 24/32 produced identical selected hashes and
token counts. Increasing a fixed buffer is therefore not a monotonic packing
optimization, and the buffer should not be multiplied by DP size. The DP8
buffer-24 point was not directly measured; the table reports the measured
buffer-32 point.

### Training trade-off

Source fill measures packing utilization, not sample loss. Unselected samples
remain buffered for later batches, and padding is excluded by attention/loss
masks. At a fixed optimizer-step count, however, lower fill means fewer source
and loss-contributing tokens.

This policy deliberately accepts the observed roughly 1-2% global utilization
gap on some 512K streams rather than delaying first forward by tens of minutes.
For the short long-context runs targeted here, entering training earlier covers
that gap under a fixed wall-clock budget. Long runs requiring exact token
parity should stop by accumulated source/loss tokens rather than optimizer
steps.

### Validation

- `pytest -q tests/data/test_dataloader.py`: 34 passed, 2 skipped
- `pytest -q tests/models/test_vlm_trainer.py`: 12 passed
- `ruff check tasks tests veomni docs`: passed
- targeted `compileall`: passed
- `git diff --check`: passed
- the complete evidence is in
  `docs/usage/long_context_dyn_bsz_cold_start.md`

This PR only reduces first-micro-batch data preparation. Whole-step
pre-materialization and periodic forward-only stalls are separate problems.

## Summary by CodeRabbit

- **New Features**
  - Added an optional context-aware buffering policy for dynamic batching.
  - Added configurable data modality support for text, multimodal, and diffusion training.
  - Applied resolved buffer sizing consistently across dataloader options.

- **Documentation**
  - Added usage guidance, configuration examples, performance measurements, and trade-offs for long-context cold starts.
  - Added the new guide to the documentation table of contents.

- **Bug Fixes**
  - Improved validation for unsupported configurations and missing dynamic-batching arguments.
  - Preserved legacy fixed-buffer behavior and plugin compatibility.

### CODE REVIEW

File: veomni/trainer/base.py [hunk 1]
```diff
@@ -505,6 +505,7 @@ def _build_dataloader(self):
         dataloader_kwargs = asdict(args.data.dataloader)
         dataloader_type = dataloader_kwargs.pop("type")
         dataloader_kwargs.pop("use_background_prefetcher", None)
+        data_modality = "multimodal" if hasattr(args.data, "mm_configs") else "text"
```
[the remark this answers is not in the document]
Comment (YLlllllllllll) [reply veomni/trainer/base.py]: Good point. Addressed in d7137ecc: DataArguments now declares an explicit ClassVar data_modality, VLM and DiT subclasses override it, and Base/BaseRL/DiT pass that property instead of relying on hasattr or hard-coded call-site values.

File: veomni/data/data_loader.py [hunk 2]
```diff
@@ -38,8 +38,61 @@
 DATALOADER_REGISTRY = Registry("dataloader")
 logger = logging.get_logger(__name__)
 
+_CONTEXT_AWARE_DYN_BSZ_BUFFER_SIZES = {
+    512 * 1024: 8,
+    1024 * 1024: 24,
+}
```
[the remark this answers is not in the document]
Comment (YLlllllllllll) [reply veomni/data/data_loader.py:49]: Thanks. This PR intentionally keeps the policy opt-in and limited to the two measured token budgets rather than extrapolating from context length alone. The shuffle-seed evidence shows that ordered-greedy fill depends on the candidate-length distribution, and larger fixed pools are not monotonic (DP2 16/24 and DP4 24/32 selected identical batches). Unmeasured contexts therefore retain the existing fixed policy until they have evidence or a separate fill-aware packing design.

**Comment by YLlllllllllll** [comment]
@Luosuu All CI checks passed on `d7137ec` (2 green) — ready to merge when you have a moment.

**Comment by YLlllllllllll** [comment]
@FoolPlayer All CI checks passed on `d7137ec`; adding you as the second default maintainer reviewer for this and future Open-VeOmni PRs.

**Comment by yl5163** [comment]
> @FoolPlayer All CI checks passed on `d7137ec`; adding you as the second default maintainer reviewer for this and future Open-VeOmni PRs.

Codex's grammar isn't that good.

File: veomni/data/data_loader.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (YLlllllllllll) [reply veomni/data/data_loader.py:49]: @cursor Thanks for calling this out. We intentionally do not generalize this policy from two measured points. The value 24 is an opt-in, per-DP-rank minimum readiness gate for the measured main-process, total-token, MBS=1 text path; it is not a buffer capacity and the existing token-budget packing remains unchanged. For 256K, 768K, 2M, and other unmeasured lengths we keep fixed/200 and fail closed because candidate-length distributions and ordered-greedy selection are not predictable from context length alone. Our DP and shuffle-seed measurements also showed non-monotonic behavior: DP2 buffers 16 and 24, and DP4 buffers 24 and 32, produced identical selected sequences and token counts. The smaller gate is intentionally used beyond the first micro-batch: the 200-candidate readiness gate is evaluated again for later micro-batches, and at 1M it can impose roughly 80 minutes of preparation time, so limiting 24 to warm-up would leave the recurring delay in place. We accept the measured roughly 1-2% utilization trade-off for this opt-in policy, while fixed/200 jobs remain unchanged. The documentation records that the evidence is from the first micro-batch and CPU data path, and recommends token-based stopping when exact parity is required. We consider the fail-closed scope an intentional design boundary and will not add unvalidated context-length rows or a generic heuristic in this PR.

**Comment by YLlllllllllll** [comment]
@Luosuu @cursor Please review the current head `ab8b0734f`.

This branch is synced with `main@ea1e68da3` and the remaining implementation issues are fixed:

- all production trainer paths now forward the runtime, policy, and explicit data modality; Omni/VLM/diffusion paths therefore fail closed for this text-only policy;
- policy resolution happens exactly once based on the actual registered builder (including native aliases and local `native` overrides), and missing required arguments now raise an actionable `ValueError` rather than a bare `KeyError`;
- new regression coverage exercises Omni forwarding, native/external dispatch, registry alias/override behavior, missing arguments, all fail-closed branches, DP1/2/4/8 policy behavior, and the minimum-gate-not-capacity invariant.

The permanent opt-in value of 24 is intentional. Switching back to 200 after warm-up would reintroduce the 200-candidate refill delay (about 80 minutes in the measured 1M path), so it would not solve the training-time stall. The default remains fixed/200, and unmeasured context lengths or unsupported runtime/count/modality/MBS combinations fail closed instead of extrapolating a heuristic. The open table thread records this design boundary and the first-micro-batch evidence limitation.

Local validation: `tests/data/test_dataloader.py` is `45 passed`; targeted Ruff/format/diff checks passed. GitHub checks are `3/3` green (auto-label, CLA, Read the Docs).

Could you please take another look?

**Comment by Luosuu** [comment]
@LiuzcEECS @Coach257 please take a look

**Comment by YLlllllllllll** [comment]
Merged current `main` at `bdc89e4d8` so this PR is no longer dirty.

`#1082` removed SeedOmni V1, so this update:
- drops `tasks/omni/train_omni_model.py` instead of resurrecting it
- keeps the buffer contract: default `200`, explicit opt-in `24` only for the measured 512K/1M main-process total-token MBS=1 text path
- retargets the scope tests to `VLMMDataArguments` / `BaseTrainer._build_dataloader`

@Luosuu @LiuzcEECS @Coach257 @cursor please re-review this head.

File: veomni/data/data_loader.py [hunk 3]
```diff
@@ -43,9 +43,73 @@
...
+            kwargs["dyn_bsz_buffer_size"] = resolve_dyn_bsz_buffer_size(
+                buffer_size=kwargs.get("dyn_bsz_buffer_size", 200),
+                buffer_policy=kwargs.get("dyn_bsz_buffer_policy", "fixed"),
+                max_seq_len=kwargs["max_seq_len"],
+                micro_batch_size=kwargs["micro_batch_size"],
+                runtime=kwargs.get("dyn_bsz_runtime", "main"),
+                count_mode=kwargs.get("dyn_bsz_count_mode", "total"),
+                data_modality=kwargs.get("data_modality", "text"),
+            )
+    return builder(**kwargs)
```
[the remark this answers is not in the document]
Comment (YLlllllllllll) [reply veomni/data/data_loader.py:124]: Fixed on `0b61e05e0`. `build_dataloader` now rejects an explicit `dyn_bsz=False` + `context_aware` combination before registry dispatch, and `build_native_dataloader` applies the same check for direct native calls. DiT `_setup` still forces `dyn_bsz=False`; if a job also sets `dyn_bsz_buffer_policy='context_aware'`, `_build_dataloader` now fails closed instead of silently using a fixed batch. Added regressions for dispatch, direct native, and the DiT path.

**Comment by YLlllllllllll** [comment]
Addressed the CodeRabbit finding on `0b61e05e0`: `dyn_bsz_buffer_policy='context_aware'` is now rejected when `dyn_bsz=False` (dispatch + native + DiT `_build_dataloader` after `_setup` disables dyn-bsz). Please re-review this head.
